packages/api-server/src/db/knex.ts

```diff
@@ -0,0 +1,315 @@
+import Knex from "knex";
+import { Store } from "../cache/store";
+import { logger } from "../base/logger";
+import {
+  TIP_BLOCK_HASH_CACHE_KEY,
+  QUERY_CACHE_EXPIRED_TIME_MS,
+} from "../cache/constant";
+
+// init cache
+const cacheStore: Store = new Store(true, QUERY_CACHE_EXPIRED_TIME_MS);
+
+const MAX_CACHE_SIZE = 100 * 1024; // 100k
+
+declare module "knex" {
+  namespace Knex {
+    interface QueryInterface<TRecord extends {} = any, TResult = any> {
+      cache: Select<TRecord, TResult>;
+    }
+  }
+}
+
+Knex.QueryBuilder.extend("cache", function useCache(this) {
+  return getCacheKey(this).then((cacheKey) => {
+    return cacheStore.get(cacheKey).then((value) => {
+      // use cache
+      if (value != null) {
+        const data = JSON.parse(value);
+        return deserializeData(data);
+      }
+
+      // query from db
+      return this.then((data) => {
+        if (data == null) {
+          return data;
+        }
+
+        // limit cache size
+        const size = Buffer.from(JSON.stringify(data), "utf-8").byteLength;
+        if (size > MAX_CACHE_SIZE) {
+          logger.debug(
+            `Max cache size(${MAX_CACHE_SIZE} bytes) exceed, total ${size} bytes.`
+          );
+          return data;
+        }
+
+        let sData;
+        try {
+          sData == serializeData(data);
+          cacheStore.insert(cacheKey, JSON.stringify(serializeData(data)));
```

Comment:
Would be good to comment why serialize before `JSON.stringify`.

Reply:
updated